### PR TITLE
Lazily generate native type layout information

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Currently, the native layout information for every non-auto-layout type is generated at type-load time, even if managed code never uses the native type. This design has lead to a few limitations in our field marshaling validation, as well as forced field marshalling initialization code to run during method-table building, causing some odd initialization ordering designs because of what data is available at what time.

This PR changes native layout information to only ever be initialized when user code or runtime code requests the native layout of a type and the type is non-blittable. To enable this, the function `IsFieldBlittable` does a fast, metadata-based blittability check that doesn't need to run all of the marshaler classification logic. When the runtime requests native layout information from a MethodTable that is blittable, it falls back to the managed information. In other cases, it generates the native information and stores it in its `LayoutEEClass` for later use.

### CODE REVIEW

File: src/vm/mlinfo.cpp [hunk 1]
```diff
@@ -5019,7 +5019,10 @@ void ArrayMarshalInfo::InitElementInfo(CorNativeType arrayNativeType, MarshalInf
 {
     CONTRACT_VOID
     {
-        STANDARD_VM_CHECK;
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
```
Comment (jkotas) [inline src/vm/mlinfo.cpp]: Why did you need to change these to `MODE_ANY`?

I think you should keep these as `STANDARD_VM_CHECK` and switch to preemptive mode around the entrypoint to the field loader.

It is a GC deadlock bug farm to have large amounts of code run under `MODE_ANY`.

File: src/vm/mlinfo.cpp [hunk 2]
```diff
@@ -1433,7 +1432,14 @@ MarshalInfo::MarshalInfo(Module* pModule,
 #endif
 )
 {
-    STANDARD_VM_CONTRACT;
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
```
Comment (jkotas) [inline src/vm/mlinfo.cpp]: Keep this as `STANDARD_VM_CONTRACT`.

Comment (jkotas) [reply src/vm/mlinfo.cpp]: (I will not comment on all places.)

File: src/vm/methodtable.cpp [hunk 3]
```diff
@@ -9955,6 +9947,35 @@ NOINLINE BYTE *MethodTable::GetLoaderAllocatorObjectForGC()
     return retVal;
 }
 
+UINT32 MethodTable::GetNativeSize()
+{
+    LIMITED_METHOD_CONTRACT;
```
Comment (jkotas) [inline src/vm/methodtable.cpp]: These methods should not be `LIMITED_METHOD_CONTRACT`. They do a lot of work inside `EnsureNativeLayoutInfoInitialized`. 

Ideally, these would be `STANDARD_VM_CONTRACT`. If it is too hard to make the callers switch to preemptive mode, these should be `THROWS; MODE_ANY`.

File: src/vm/mlinfo.cpp (hunk 1 above)
Comment (jkoritzinsky) [reply src/vm/mlinfo.cpp]: I believe I changed them to avoid needing to do GC-mode switching in `marshalnative.cpp` since it seemed that everything they called was also `MODE_ANY`. I'll change `marshalnative.cpp` to switch to preemptive GC when calling into the interop subsystem.

File: src/vm/class.cpp [hunk 4]
```diff
@@ -4024,43 +3770,528 @@ void EEClassLayoutInfo::ParseFieldNativeTypes(
...
+            COMPlusThrow(kTypeLoadException, IDS_CANNOT_MARSHAL_RECURSIVE_DEF, GetFullyQualifiedNameForClassW(pMT));
+        }
+
+        if (pClass->GetNativeLayoutInfo() == nullptr)
+        {
+            EEClassNativeLayoutInfo* pNativeLayoutInfo;
+            CollectNativeLayoutFieldMetadataThrowing(pMT, &pNativeLayoutInfo);
+            ((LayoutEEClass*)pClass)->m_nativeLayoutInfo.SetValue(pNativeLayoutInfo);
+#ifdef FEATURE_HFA
+            pMT->GetClass()->CheckForNativeHFA();
```
Comment (jkotas) [inline src/vm/class.cpp]: This does not look thread safe. I think all initialization of `pNativeLayoutInfo` needs to be done before `m_nativeLayoutInfo` is set to point to it.

File: src/vm/fieldmarshaler.h [hunk 5]
```diff
@@ -212,17 +216,234 @@ class NativeFieldDescriptor
...
+    }
+#else
+    bool IsNativeStructPassedInRegisters() const
+    {
+        return false;
+    }
+#endif
 };
 
+inline CorElementType EEClassNativeLayoutInfo::GetNativeHFATypeRaw() const
```
Comment (jkotas) [inline src/vm/fieldmarshaler.h]: This does not look like a function that will ever inline.

File: src/vm/class.cpp [hunk 6]
```diff
@@ -4024,43 +3770,528 @@ void EEClassLayoutInfo::ParseFieldNativeTypes(
...
+// Called from the clsloader to load up and summarize the field metadata
+// for layout classes.
+//
+// Warning: This function can load other classes (esp. for nested structs.)
+//=======================================================================
+#ifdef _PREFAST_
+#pragma warning(push)
+#pragma warning(disable:21000) // Suppress PREFast warning about overly large function
+#endif
+VOID EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing(
```
Comment (jkotas) [inline src/vm/class.cpp]: Make a new file for this and move all EEClassLayoutInfo stuff to it?

File: src/vm/class.cpp [hunk 7]
```diff
@@ -4024,43 +3770,528 @@ void EEClassLayoutInfo::ParseFieldNativeTypes(
...
-
 }
+
+//=======================================================================
+// Called from the clsloader to load up and summarize the field metadata
+// for layout classes.
+//
+// Warning: This function can load other classes (esp. for nested structs.)
+//=======================================================================
+#ifdef _PREFAST_
```
Comment (jkotas) [inline src/vm/class.cpp]: You can delete these "Suppress PREFast warning about overly large function" on the functions that you are touching.

Comment (jkotas) [reply src/vm/class.cpp]: (We do not run the analyzer to check for these anymore. And the functions are not that big anyway thanks to your refactoring.)

File: src/vm/classlayoutinfo.cpp [hunk 8]
```diff
@@ -0,0 +1,1105 @@
...
+            CollectNativeLayoutFieldMetadataThrowing(pMT, &pNativeLayoutInfo);
+#ifdef FEATURE_HFA
+            CheckForNativeHFA(pMT, pNativeLayoutInfo);
+#endif
+#ifdef UNIX_AMD64_ABI
+            SystemVAmd64CheckForPassNativeStructInRegister(pMT, pNativeLayoutInfo);
+#endif
+
+            ((LayoutEEClass*)pClass)->m_nativeLayoutInfo.SetValue(pNativeLayoutInfo);
+            if(IsStructMarshalable(pMT))
```
Comment (jkotas) [inline src/vm/classlayoutinfo.cpp]: This still look thread unsafe.

File: src/vm/class.cpp [hunk 9]
```diff
@@ -1463,134 +1473,33 @@ EEClass::CheckForHFA()
...
-    return hfaType;
-}
-
 #ifdef FEATURE_HFA
 //
 // The managed and unmanaged views of the types can differ for non-blitable types. This method
 // mirrors the HFA type computation for the unmanaged view.
 //
-VOID EEClass::CheckForNativeHFA()
+void CheckForNativeHFA(MethodTable* pMT, EEClassNativeLayoutInfo* pNativeLayoutInfo)
```
Comment (jkotas) [inline src/vm/class.cpp]: Move to classlayoutinfo.cpp?

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.
